### PR TITLE
Add implicit reparameterization gradient to tfd.TwoPieceNormal

### PR DESCRIPTION
**Pull request by leandrolcampos**
Add implicit reparameterization gradient w.r.t. `skewness`, making the distribution `tfd.TwoPieceNormal` fully reparameterized. The applied approach to compute the implicit gradient was proposed by Figurnov, Mohamed, and Mnih (2018)[1].

In addition to implementing the implicit gradient described above, this PR makes the following changes:
* In the file `tensorflow_probability/python/distributions/BUILD`:
    * Add a new dependency
* In the file `tensorflow_probability/python/distributions/two_piece_normal.py`:
    * Reorganize the file, putting all functions below the class definition
    * Run all Tensor args through `tf.convert_to_tensor` in the functions `standardize`, `cdf`, and `quantile`, making them consistent with `random_two_piece_normal`
* In the file `tensorflow_probability/python/distributions/two_piece_normal_test.py`:
    * Improve `testSample`, testing different sample shapes
    * Add `testDifferentiableSampleNumerically` and `testDifferentiableSampleAnalytically`, testing the gradients of the samples w.r.t. `loc`, `scale`, and `skewness`

#### Reference

[1] Michael Figurnov, Shakir Mohamed, and Andriy Mnih. Implicit Reparameterization Gradients. In _Advances in Neural Information Processing Systems_, 31, 2018.
https://arxiv.org/abs/1805.08498

### CODE REVIEW

**Comment by SiegeLordEx** [comment]
Thanks for this. I'll take a look at this by EOW.

File: tensorflow_probability/python/distributions/two_piece_normal.py [hunk 1]
```diff
@@ -490,3 +393,301 @@ def _parameter_control_dependencies(self, is_init):
...
+
+def _two_piece_normal_sample_bwd(_, aux, dy):
+  """The gradients of Two Piece Normal samples w.r.t. skewness."""
+  skewness, samples = aux
+  broadcast_skewness = tf.broadcast_to(skewness, ps.shape(samples))
+
+  grad = dy * _two_piece_normal_sample_gradient(broadcast_skewness, samples)
+  # Sum over the sample dimensions. Assume that they are always the first
+  # ones.
+  num_sample_dimensions = (tf.rank(broadcast_skewness) -
```
Comment (SiegeLordEx) [inline tensorflow_probability/python/distributions/two_piece_normal.py:604]: In this function, you should use `ps.rank` and `ps.range`, as you're dealing with shapes and we want those to be statically propagated if possible.

File: tensorflow_probability/python/distributions/two_piece_normal_test.py [hunk 2]
```diff
@@ -348,6 +350,55 @@ def sampler(loc, scale, skewness):
...
+    """Test the gradients of the samples w.r.t. skewness."""
+    sample_shape = [int(2e5)]
+    seed = test_util.test_seed()
+
+    def sampler(skewness):
+      loc = tf.constant(0., self.dtype)
+      scale = tf.constant(1., self.dtype)
+      dist = tfd.TwoPieceNormal(
+          loc, scale=scale, skewness=skewness, validate_args=True)
+      return tf.reduce_mean(dist.sample(sample_shape, seed=seed))
```
Comment (SiegeLordEx) [inline tensorflow_probability/python/distributions/two_piece_normal_test.py:364]: Can you add some non-linearity after the mean computation (tanh or something), to make sure this is checking the that the backwards pass gradients are correctly incorporated.

File: tensorflow_probability/python/distributions/two_piece_normal_test.py [hunk 3]
```diff
@@ -348,6 +350,55 @@ def sampler(loc, scale, skewness):
...
+        np.linspace(0.1, stop=10., num=n_params), dtype=self.dtype)
+    skewness = tf.constant(
+        np.linspace(0.75, stop=1.33, num=n_params), dtype=self.dtype)
+
+    seed = test_util.test_seed()
+
+    def sampler(loc, scale):
+      dist = tfd.TwoPieceNormal(
+          loc=loc, scale=scale, skewness=skewness, validate_args=True)
+      return dist.sample(sample_shape, seed=seed)
```
Comment (SiegeLordEx) [inline tensorflow_probability/python/distributions/two_piece_normal_test.py:391]: Same non-linearity suggestion here as well (although here you want something that's easy to analytically handle).

**Review by SiegeLordEx — CHANGES_REQUESTED** [review]
This looks excellent. Just a few minor suggestions.

File: tensorflow_probability/python/distributions/two_piece_normal.py (hunk 1 above)
Comment (leandrolcampos) [reply tensorflow_probability/python/distributions/two_piece_normal.py]: Done.

File: tensorflow_probability/python/distributions/two_piece_normal_test.py (hunk 2 above)
Comment (leandrolcampos) [reply tensorflow_probability/python/distributions/two_piece_normal_test.py]: As in `truncated_normal_test.testReparametrizable`, I used the nonlinear function `tf.abs` inside `tf.reduce_mean`. Is this OK?

File: tensorflow_probability/python/distributions/two_piece_normal_test.py (hunk 3 above)
Comment (leandrolcampos) [reply tensorflow_probability/python/distributions/two_piece_normal_test.py]: Here I used the nonlinear function `tf.math.exp`.

File: tensorflow_probability/python/distributions/two_piece_normal_test.py (hunk 2 above)
Comment (SiegeLordEx) [reply tensorflow_probability/python/distributions/two_piece_normal_test.py]: I think it should be okay. The goal is to get `dy` to not always be 1 and I think abs flips it to -1 sometimes?

**Comment by leandrolcampos** [comment]
Thanks for your suggestions, @SiegeLordEx! It's a pleasure to learn from you!

**Review by SiegeLordEx — APPROVED** [review]

**Comment by SiegeLordEx** [comment]
Thanks for the contribution! I'll send it along to our internal system now, as before.

**Comment by SiegeLordEx** [comment]
In internal review we focused on the numerical stability of the gradient implementation, and we made the following changes:

- After a bit of expansion and simplification, we reduced the left and right gradient expressions to be expressed in terms of the Mills ratio, which we could compute accurately via `erfcx` special function.
- We applied the double-where trick to avoid NaNs in the second derivative (see https://github.com/tensorflow/probability/blob/main/discussion/where-nan.pdf)

**Comment by leandrolcampos** [comment]
> In internal review we focused on the numerical stability of the gradient implementation, and we made the following changes:
> 
> * After a bit of expansion and simplification, we reduced the left and right gradient expressions to be expressed in terms of the Mills ratio, which we could compute accurately via `erfcx` special function.
> * We applied the double-where trick to avoid NaNs in the second derivative (see https://github.com/tensorflow/probability/blob/main/discussion/where-nan.pdf)

Hi @SiegeLordEx,

Thank you so much for the improvements in the gradient implementation. I'll study the changes you made and the material you shared about the double-where trick. As I said before, it's a pleasure to learn from you and contribute to TensorFlow Probability!
